Add scr_poststage

tonyhutter opened the pull request:
This adds in the `scr_poststage `script and all associated changes.  This allows a user to start a checkpoint transfer right before their job ends and have the transfer continue "in the background" after the job. The checkpoint files are then finalized using the `scr_poststage` script after the transfer ends.  All this uses the IBM Burst Buffer software.

Specific changes:

- Added `scr_flush_file -r` to resume and finalize a transfer.
- Added `scr_flush_file -s` to write a summary file for a transfer that has been resumed and finalized.
- Added `examples/test_scr_poststage.sh` for testing `scr_poststage.sh`
- Added a `XFER=BBAPI_POSTSTAGE` transfer type to the storage description section in `scr.conf`.  This tells SCR to start the transfers but let scr_poststage.sh do the finalizing.
- Update documentation
- Add `scr_poststage.sh`

tonyhutter commented:
@adammoody note that I did try implementing this under "async" mode, but I noticed it would still eventually wait for the files to finish, which isn't what we want.  That's why I ultimately added the `XFER=BBAPI_POSTSTAGE` storage descriptor transfer type.  That transfer type tells SCR to "blindly start the transfer and let scr_poststage deal with it".

Another note: the way I implemented this will finalize all the un-finalized bbapi_poststage transfers.  That means if you spawn off multiple BBAPI_POSTSTAGE checkpoint transfers in parallel, they will all get finalized in post stage.  One thing we may want to do later is to only have one post-stage transfer going at any given time.  So if you launched one transfer, and then an hour later launched another one, the 2nd transfer would cancel the first transfer.  That would save BB <-> GPFS bandwidth since it would only transfer the newest checkpoint to GPFS.

adammoody commented on src/scr_index.c:
```diff
@@ -2167,11 +2168,6 @@ int index_add(const spath* prefix, const char* name)
               int complete = 0;
               kvtree_util_get_int(summary, SCR_SUMMARY_6_KEY_COMPLETE, &complete);
 
-              /* inform user that we found the dataset */
-              char* path = spath_strdup(dataset_path);
-              printf("Found `%s' as dataset %d at %s\n", name, id, path);
-              scr_free(&path);
-
```
I'd like to keep this in there.  Was it getting in the way?

adammoody commented on src/scr_util_mpi.c:
```diff
@@ -352,7 +355,7 @@ int scr_axl(
       scr_err("Failed to add file to AXL transfer handle %d: %s --> %s @ %s:%d",
         id, src_file, dest_file, __FILE__, __LINE__
       );
-      rc = SCR_FAILURE;
+      return SCR_FAILURE;
```
I think we need to keep this with a set rather than a return to avoid deadlock below.

tonyhutter replied:
Not sure I follow - which deadlock are you referring to?  Do we need to still call `AXL_Free_comm()` on failures?

adammoody replied:
The particular deadlock I'm worried about here happens if ``AXL_Add`` fails on one process, but succeeds on other procs.  In that case, that process would now return with this change whereas the other procs will go on to call ``AXL_Dispatch_comm``, but that contains an allreduce, which leads to the deadlock.

On the other hand, the existing code (before your change here) was also broken.  We shouldn't call ``AXL_Dispatch`` if ``AXL_Add`` fails.  I think we need an allreduce inbetween to check that ``AXL_Add`` succeeded everywhere.  If not we need to call ``AXL_Free`` and return everywhere with an error.  I think a second problem in the existing code is that we shouldn't call ``AXL_Wait`` if ``AXL_Dispatch_comm`` returns an error.  There may be some other related problems with the existing code.

One option might be to drop these changes from the PR and open a separate issue to clean all of this up in a separate PR.

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,167 @@
+    *        ckpt.1/rank_74.ckpt
+    *...
+    */
+    sprintf(num_str, "%lu", i);
+    rank_subtree =  kvtree_get(ranks_tree, num_str);
```
Let's fuse this if getf works and then drop num_str:
```suggestion
    rank_subtree =  kvtree_getf(ranks_tree, "%lu", i);
```

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,167 @@
+      scr_err("Couldn't get RANK subtree, rc = %d\n", rc);
+      goto out;
+    }
+    file = NULL;
+    rc = kvtree_util_get_str(rank_subtree, "FILE", &file);
+    if (rc != KVTREE_SUCCESS) {
+      scr_err("Couldn't get file name to resume, rc = %d\n", rc);
+      goto out;
+    }
```
We're not using the ``file`` afterwards.  Is this just a check that a rank has a file listed?

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,167 @@
+
+  rc = kvtree_read_scatter_single(rank2file_path, ranks_tree);
+  if (rc) {
+    scr_err("kvtree_read_scatter_single(%s) = %d\n", __func__, rank2file_path, rc);
```
I don't think we need ``\n`` on scr_err calls.  If I'm remembering correctly, let's remove those.
```suggestion
    scr_err("kvtree_read_scatter_single(%s) = %d", __func__, rank2file_path, rc);
```

adammoody replied:
For scr_err messages, let's also update those to end with ``@ %s:%d`` and print __FILE__ and __LINE__ values like this example:
```
      scr_err("Failed to read SCR_FLUSH_ASYNC_BW successfully @ %s:%d",
        __FILE__, __LINE__
      );
```

adammoody replied:
And let's double check each of those that the printf format string matches the type and order of the arguments.  For example, I think this particular line is missing a ``%s`` somewhere or we should drop ``__func__``.
```
scr_err("kvtree_read_scatter_single(%s) = %d\n", __func__, rank2file_path, rc);
```
Those lines need extra manual care, since the compiler often can't catch them, and they don't execute at runtime since they are error messages.

tonyhutter replied:
That looks like a debug print that snuck in, I'll remove it.  I'll change the other ones I added to include __FILE__ and __LINE__ if they don't have them already.

adammoody commented:
Thanks @tonyhutter .  This will be a really nice addition for users.  A couple of high-level items:

Instead of finalizing *all* transfers in poststage, we only want that to do that on the final transfer, i.e., the one that SCR would normally wait on during SCR_Finalize.
https://github.com/LLNL/scr/blob/c53e3f7bdce2cf19aa73d87f9b2c7d651ee133a9/src/scr.c#L2374
We need SCR to complete all of the others during the run, because SCR needs to know when the transfer has completed, so that it can delete it from cache.  In a long run, the app might transfer 100 datasets during the course of the run, but it might only have enough space to hold one dataset in cache at a time.

We'll want to add this to the async flush path, too.

For the sync flush path, we want SCR to wait as it currently does.  When configured for sync flush, the user expects the dataset to be on the parallel file system when SCR returns from sync flush operation.  Though, there are still two spots where we can use this with poststage even with a sync flush:
1) if SCR starts a sync flush, but then the job allocation terminates before the transfer finished, we could still finalize it during the poststage.
2) as a configuration option for SCR_Finalize to just initiate the transfer and exit rather than wait

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,76 @@
+# Finalize each dataset listed in `scr_index -l -p` that is NOT
+# flushed.
+#
+# $ scr_index -l -p $PREFIX
+# DSET VALID FLUSHED             CUR NAME
+#   2 NO                          ckpt.2
+#   1 NO                          ckpt.1
```
Let's use the flush file to find the set of datasets to finish, rather than the index file.  The flush file should only contain info for datasets that have been flushed in the current job.  The index file will also contain info for lots of older datasets written from earlier jobs.

adammoody replied:
There are a lot of similarities in what we're doing in poststage with what we do in our scavenge operation.  As future work, we'll need to integrate poststage capability with scavenge.  That may be a lot of work, so we'll save that poststage+scavenge step for a future PR.

In the meantime, the logic to identify the datasets to scavenged is here:
https://github.com/LLNL/scr/blob/develop/scripts/common/scr_postrun.in#L112-L249

There are two loops there.  The goal of that code is to scavenge every dataset that is marked as output and also the most recent checkpoint or at least the most recent checkpoint that comes before the first failed dataset.

We could use similar scr_flush_file commands to get the dataset ids and names to wait on in the poststage script.  We are likely missing the info about whether the dataset was flushed with BBAPI.

adammoody commented on src/scr_fetch.c:
```diff
@@ -239,9 +239,18 @@ static int scr_fetch_data(
     //const scr_storedesc* storedesc = scr_cache_get_storedesc(cindex, id);
     axl_xfer_t xfer_type = scr_xfer_str_to_axl_type(SCR_FETCH_TYPE);
+    int no_wait = 0;
+    if (strcasecmp(SCR_FETCH_TYPE, "bbapi_poststage") == 0) {
+      /*
+       * "bbapi_poststage" means the BBAPI can continue transferring the
+       * checkpoints "in the background" after your job ends.  The transfers
+       * will then ultimately get finalized by the scr_poststage script.
+       */
+      no_wait = 1;
+    }
 
     /* fetch these files into the directory */
-    if (scr_axl(dset_name, num_files, src_filelist, dest_filelist, xfer_type, scr_comm_world) != SCR_SUCCESS) {
+    if (scr_axl(dset_name, NULL, num_files, src_filelist, dest_filelist, xfer_type, no_wait, scr_comm_world) != SCR_SUCCESS) {
```
We still want the fetch to wait for sure.  This is where SCR reads data from the file system and writes it to cache (opposite of flush).  There is a potential BB change here using prestage, but fetch won't change for poststage.

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,167 @@
+    scr_err("Error: AXL_Finalize(), = %d\n", id, rc);
+  }
+end:
+  return rc;
```
We should double check the return value here for cases above.  I'm guessing we at least want to convert from AXL return codes to something else.  Or we could have it return void since we're not using that from where it's called.

tonyhutter replied:
I'll just make it a `void` and simplify the function.

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,167 @@
+    */
+    sprintf(num_str, "%lu", i);
+    rank_subtree =  kvtree_get(ranks_tree, num_str);
+    if (!rank_subtree) {
+      scr_err("Couldn't get RANK subtree, rc = %d\n", rc);
+      goto out;
+    }
```
There are datasets where some ranks contribute 0 files.  I forget how that would be reflected here.  Does it end up with an empty entry, or is there no entry at all for that rank?

tonyhutter replied:
I don't know the answer, nor do I know how I would test that.  I think the easiest thing would be to to skip any ranks that don't have a subtree or a `FILE` entry, and don't report an error.

tonyhutter replied:
> I don't know the answer, nor do I know how I would test that. 

Duh, my mistake! ... I missed that already gave the instructions on how to test it:
https://github.com/LLNL/scr/pull/316#issuecomment-830471759

I'll give it a try and report back

tonyhutter replied:
Tested, and it looks like this:
```
--- ranks_tree ---
      13
        FILE
          timestep.11/rank_13.0.ckpt
      12
      15
        FILE
          timestep.11/rank_15.0.ckpt
          timestep.11/rank_15.1.ckpt
          timestep.11/rank_15.2.ckpt
      14
        FILE
          timestep.11/rank_14.0.ckpt
          timestep.11/rank_14.1.ckpt
...
```
So we can just skip the entries that have no `FILE`

adammoody commented on src/scr_flush_sync.c:
```diff
@@ -59,6 +61,10 @@ static int scr_flush_sync_data(scr_cache_index* cindex, int id, kvtree* file_lis
   }
   MPI_Barrier(scr_comm_world);
 
+  dataset_path_str = spath_strdup(dataset_path);
+  asprintf(&state_file, "%s/rank_%d.state_file", dataset_path_str, scr_my_rank_world);
+  scr_free(&dataset_path_str);
```
Let's change this to use spath, even though it adds a line:
```suggestion
  spath* state_file_path = spath_dup(dataset_path);
  spath_append_strf(state_file_path, "rank_%d.state_file", scr_my_rank_world);
  char* state_file = spath_strdup(state_file_path);
  scr_path_delete(&state_file_path);
```

tonyhutter commented:
Thanks for reviewing this.  I'll need some time to look over the fixes.

At a high level - could we get the behavior you're describing by having any new BBAPI_POSTSTAGE transfer cancel any existing BBAPI_POSTSTAGE transfers before starting?  That way you'll only have one poststage transfer in flight at a time, and it will always be for the newest checkpoint.  Assume it does this both sync and async.  Would that work, or does it fall over somehow?

adammoody commented:
The cancel is could be a nice optimization, but it doesn't solve the problem in general.  The transfers could be output datasets, meaning the user does actually need each one saved on the parallel file system.

adammoody commented:
Another thing to consider for testing.  One can configure a run to write some datasets to /dev/shm and others to BBPATH.  For async flushes, the user would then configure SCR to flush /dev/shm with something like PTHREAD and use BBAPI for BBPATH.  In poststage, we need to correctly handle both types.  It may just work as is, but if not, we might also want to record the transfer type in the flush file.

adammoody commented:
The usage model we want to add first is to allow the user to configure their run to either wait or not wait in ``SCR_Finalize()`` on a BBAPI flush for both sync and async flushes:
```
if (poststage finalize is enabled and last checkpoint is BBAPI flushable) {
  if (async flush of last checkpoint has not already been started) {
    // start BBAPI flush, but don't wait on it
    scr_flush_async_start()
  } else {
    // BBAPI flush of last checkpoint started before SCR_Finalize
    // will be completed in poststage, so nothing else to do
  }
} else {
 // either poststage finalize is not enabled, or it is enabled but last checkpoint can't use BBAPI
 // use SCR's existing flush_sync() or flush_async_wait()
}
```
Additionally, we want to complete any BBAPI flushes that were started but not completed in case the code failed to call ``SCR_Finalize()``.  The code may have run out of time, or it may have encountered a failure after it started the BBAPI flush.

adammoody commented:
The existing SCR flush logic in SCR_Finalize is here:
https://github.com/LLNL/scr/blob/c53e3f7bdce2cf19aa73d87f9b2c7d651ee133a9/src/scr.c#L2326-L2371

adammoody commented:
To test a config using async flushes from both /dev/shm and BBAPI, a config file like this might work (not tested):
```
SCR_COPY_TYPE=FILE
SCR_FLUSH=1
SCR_FLUSH_ASYNC=1

STORE=/dev/shm GROUP=NODE COUNT=1 FLUSH=PTHREAD
STORE=$BBPATH  GROUP=NODE COUNT=3 FLUSH=BBAPI
 
CKPT=0 INTERVAL=1 STORE=/dev/shm GROUP=NODE TYPE=XOR SET_SIZE=16
CKPT=1 INTERVAL=2 STORE=$BBPATH  GROUP=NODE TYPE=XOR SET_SIZE=16
```

adammoody commented:
To test an SCR job where some ranks don't contribute any files to the dataset, one can use ``test_api_multiple``.
```
srun -n4 -N4 ./test_api_multiple
```
With that configuration, each rank writes the following number of files for each dataset:
```
rank 0: 0 files
rank 1: 1 file
rank 2: 2 files
rank 3: 3 files
```
I think one needs to run that with at least two ranks.  The file count wraps back to 0 every 4 ranks, so if one ran with at least 5 procs, rank 4 would also write 0 files.
https://github.com/LLNL/scr/blob/c53e3f7bdce2cf19aa73d87f9b2c7d651ee133a9/examples/test_api_multiple.c#L75

adammoody commented:
While at it, other options in ``test_api`` might be useful.  A more complicated run:

```
srun -n4 -N4 ./test_api --times 10 --output 5 --size 512MB --seconds 30
```

If I remember correctly, this will write 10 datasets in total.  Every 5th dataset will be a pure output dataset (not a checkpoint).  Each file will be ~512 MB in size.  It will sleep for 30 seconds between writing consecutive datasets.

I don't think ``test_api_multiple`` supports all of those.

adammoody commented on examples/test_scr_poststage.sh:
```diff
@@ -0,0 +1,78 @@
+    echo "
+SCR_COPY_TYPE=FILE
+
+SCR_CLUSTER_NAME=`hostname`
+
```
```suggestion
```

adammoody replied:
This isn't required anymore, so let's drop it.

adammoody commented on examples/test_scr_poststage.sh:
```diff
@@ -0,0 +1,78 @@
+
+CKPT=0 INTERVAL=1 GROUP=NODE   STORE=$BB_DIR TYPE=PARTNER
+
+SCR_CNTL_BASE=$BB_DIR
+SCR_CACHE_BASE=$BB_DIR
+
+CNTLDIR=$BB_DIR BYTES=10MB
+CACHEDIR=$BB_DIR BYTES=10MB
```
These lines aren't necessary, and they might lead to problems down the road, e.g., if we change test_api to write files larger than 10MB.
```suggestion
```

adammoody replied:
Or are you relying on this for some reason?

tonyhutter commented:
I had a really good call with @adammoody today where we discussed the changes he had in mind for the PR, and I'm fully on-board with the changes. In a nutshell:

Get rid of the special `XFER=BBAPI_POSTSTAGE` storage descriptor value and just use BBAPI under the async flush codepath instead.  Along with this, add a `SCR_FLUSH_POSTSTAGE=1` option to tell async not to block on transfers in SCR_Finalize.

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,81 @@
+
+# Finalize each dataset listed in the flush file
+for ID in $(scr_flush_file --dir $PREFIX --list-ckpt) ; do
+	CKPT="ckpt.$ID"
```
For the name, let's use the dataset name as given by the user.  You can get that like so:
https://github.com/LLNL/scr/blob/c53e3f7bdce2cf19aa73d87f9b2c7d651ee133a9/scripts/common/scr_postrun.in#L128

tonyhutter replied:
:+1:

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,81 @@
+		continue
+	fi
+
+	echo "Setting current checkpoint to $CKPT" >> $LOGFILE
+	if ! scr_index -p $PREFIX --current=$CKPT &>> $LOGFILE ; then
+		echo "Couldn't set current checkpoint to $CKPT" >> $LOGFILE
+		continue
+	fi
```
Not all datasets are checkpoints.  I don't remember what happens if we set a non-checkpoint to be ``--current``.  This is a reminder that I need to review that.

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,81 @@
+		continue
+	fi
+
+	echo "Adding checkpoint $CKPT to index" >> $LOGFILE
```
```suggestion
	echo "Adding dataset $CKPT to index" >> $LOGFILE
```

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,81 @@
+
+	echo "Adding checkpoint $CKPT to index" >> $LOGFILE
+	if ! scr_index -p $PREFIX --add=$CKPT &>> $LOGFILE ; then
+		echo "Couldn't add checkpoint $CKPT to index" >> $LOGFILE
```
```suggestion
		echo "Couldn't add dataset $CKPT to index" >> $LOGFILE
```

adammoody commented on src/scr.c:
```diff
@@ -931,6 +931,13 @@ static int scr_get_params()
   /* specify whether to use asynchronous flush */
   if ((value = scr_param_get("SCR_FLUSH_ASYNC")) != NULL) {
     scr_flush_async = atoi(value);
+
+    /* Specify whether our flush will be finalized in poststage (currently
+     * only supported with BBAPI).
+     */
+    if ((value = scr_param_get("SCR_FLUSH_POSTSTAGE")) != NULL) {
+      scr_flush_poststage = atoi(value);
+    }
```
Let's make this generally available, and not dependent on ``SCR_FLUSH_ASYNC``.

adammoody commented on src/scr_conf.h:
```diff
@@ -220,8 +220,14 @@
 /* whether to switch from synchronous to asynchronous flushes */
 #ifndef SCR_FLUSH_ASYNC
 #define SCR_FLUSH_ASYNC (0)
+
+/* Finalize async transfers in scr_poststage.sh rather than in SCR_Finalize() */
+#ifndef SCR_FLUSH_POSTSTAGE
+#define SCR_FLUSH_POSTSTAGE (0)
 #endif
 
```
Let's similarly pull this out a level.

adammoody commented on doc/rst/users/config.rst:
```diff
@@ -358,6 +358,18 @@ The table in this section specifies the full set of SCR configuration parameters
+       `scr_poststage.sh` as your 2nd-half post-stage script in bsub to
+       finalize the transfers.  See `examples/test_scr_poststage.sh` for a
+       detailed example.
+
+       Note: SCR_FLUSH_POSTSTAGE requires SCR_FLUSH_ASYNC=1.
```
When we get around to scavenge+poststage support, this setting will be useful to know independently of the value of ``SCR_FLUSH_ASYNC``.  Let's drop this restriction to make it more general.

adammoody commented:
Thanks, @tonyhutter .  We could split the PTHREAD+BBAPI combo support into a future PR if that helps.

I'd like to see if we can verify the case where some ranks don't write any files in this PR, though.  There are existing codes that do that.

adammoody commented on src/scr.c:
```diff
@@ -2358,7 +2376,7 @@ int SCR_Finalize()
 
   /* flush checkpoint set if we need to */
-  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
+  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id) && !poststage_skip) {
```
There's one additional case we want to add here.  It's possible that SCR hasn't started the flush of the most recent checkpoint.  For example, one might set ``SCR_FLUSH > 1`` and then call ``SCR_Finalize`` immediately after a writing a checkpoint that SCR didn't initiate a flush for.  SCR currently accounts for such a checkpoint by calling ``scr_flush_sync`` here.  We want to extend this so that if they are using poststage and if that checkpoint is in BBAPI, we call ``scr_flush_async_start`` instead of ``scr_flush_sync``.

adammoody replied:
To expand on that, if ``SCR_FLUSH=2``, then on a new run, SCR would flush the following:
```
dataset id 1 = not flushed
dataset id 2 = async flush started in ``SCR_Complete_output``
dataset id 3 = not flushed
SCR_Finalize()
```
Given a pattern like that, in ``SCR_Finalize`` dataset 2 should be the ongoing async flush.  SCR would wait on dataset 2 to finish by calling ``scr_flush_async_wait`` and then call ``scr_flush_sync`` to flush dataset 3 (which happens to be the most recent checkpoint).

We'd want to initiate the flush of dataset 3 by calling ``scr_flush_async_start`` instead.

tonyhutter replied:
Thanks for pointing this out.  I'm thinking something like?:
```C
  /* flush checkpoint set if we need to */
  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
    int flush_rc;

    if (scr_my_rank_world == 0) {
      scr_dbg(2, "Begin flushing checkpoints that haven't started @ %s:%d", __FILE__, __LINE__);
    }
    if (!poststage_skip) {
      flush_rc = scr_flush_sync(scr_cindex, scr_ckpt_dset_id);
    } else { 
      flush_rc = scr_flush_async_start(scr_ckpt_dset_id);
    }
  }
  if (flush_rc != SCR_SUCCESS) {
    scr_abort(-1, "Flush of dataset %d failed @ %s:%d",
        scr_ckpt_dset_id, __FILE__, __LINE__
        );
  }
```

adammoody replied:
Yeah, that should work I think.  The double negative in ``!poststage_skip`` keeps twisting my mind, though.  Let's rename this variable to ``poststage`` and flip the meaning to be positive:
```
if (poststage) {
  scr_flush_async_start
} else {
  scr_flush_sync
}
```

adammoody replied:
When using poststage, we'll also eventually want to change things to avoid waiting on dataset 2 to finalize in this example.

adammoody replied:
Oh, I just remembered.  If also having dataset 2 finalize in the poststage script would require making more extensive changes in SCR to allow for multiple outstanding flushes at once, let's delay that extra bit to a future PR.  That may be a bigger change than we want to bite off in this PR.

tonyhutter replied:
Would that mean the `SCR_Finalize()` behavior would change to a more general "we don't wait on any async, BBAPI-post-stageable transfers?" So with your example, it would mean that `SCR_Finalize()` doesn't wait on dataset 2, and would start dataset 3, and both 2 & 3 would get finalized in `scr_poststage.sh`?

tonyhutter replied:
Ah ok, just saw your last comment after I had already posted.

tonyhutter replied:
I'll keep things as we talked about earlier:
```C
  /* flush checkpoint set if we need to */
  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
    int flush_rc;

    if (scr_my_rank_world == 0) {
      scr_dbg(2, "Begin flushing checkpoints that haven't started @ %s:%d", __FILE__, __LINE__);
    }

    if (poststage) {
      if (scr_flush_file_is_flushing(scr_ckpt_dset_id)) {
        flush_rc = SCR_SUCCESS;
      } else {
        flush_rc = scr_flush_async_start(scr_cindex, scr_ckpt_dset_id);
      }
    } else {
      /* Normal case */
      flush_rc = scr_flush_sync(scr_cindex, scr_ckpt_dset_id);
    }

    if (flush_rc != SCR_SUCCESS) {
      scr_abort(-1, "Flush of dataset %d failed @ %s:%d",
          scr_ckpt_dset_id, __FILE__, __LINE__
          );
    }
  }
```

adammoody commented on src/scr.c:
```diff
@@ -2417,11 +2435,13 @@ int SCR_Finalize()
-    scr_abort(-1, "Failed to finalize AXL library @ %s:%d",
-      __FILE__, __LINE__
-    );
+  if (!poststage_skip) {
+    int axl_rc = AXL_Finalize_comm(scr_comm_world);
+    if (axl_rc != AXL_SUCCESS) {
+      scr_abort(-1, "Failed to finalize AXL library @ %s:%d",
+	__FILE__, __LINE__
+      );
+    }
```
Does it break things if we leave this enabled?  I'm worried about introducing memory leaks if we don't call it.

tonyhutter replied:
No, looks like I don't need that at all.  I'll remove

tonyhutter commented:
@adammoody please take another look, I think I've made most of the changes from your feedback.  Some specific things that come to mind:

- Tested with both pthreads and BBAPI cache devices (below) and saw poststage work as intended.  That is, I adjusted `test_api --times` to test both devices being the "last checkpoint" cache, and saw it correctly wait on the pthreads cache device in SCR_Finalize, and not wait on it for the BB cache (which used scr_poststage.sh).
```
CKPT=0 INTERVAL=1 STORE=/dev/shm GROUP=NODE TYPE=XOR SET_SIZE=16
CKPT=1 INTERVAL=2 STORE=$BBPATH  GROUP=NODE TYPE=XOR SET_SIZE=16
```
- Updated code to work when a rank doesn't doesn't write a checkpoint (using the test_api_multiple).

- Added a lockfile to ensure only one poststage is running per prefix at a time.
- Misc other fixes you suggested

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+			continue
+		fi
+
+		echo "Finalizing transfer for dataset $ID ($DSET)" >> $LOGFILE
+		if ! scr_flush_file -r -s $ID --dir $PREFIX &>> $LOGFILE ; then
+			echo "Error: Can't resume dataset $ID, rc=$?" >> $LOGFILE
+			continue
+		fi
```
Thanks for taking a look at how pthread transfers are handled in ``SCR_Finalize``.  The other spot we want to account for a pthread transfer is here.  Perhaps the job started a pthread flush but died or ran out of time before it could be finalized within the library.  In that case, I think we'll fall through to this point.  It makes sense for the resume to fail and mark that as an error as you have.  We should also verify that ``scr_flush_file`` recognizes it failed here.

Do you know whether that works?

tonyhutter replied:
I don't know, let me give it a quick test and report back

tonyhutter replied:
Just ran the test - the good news is that an interrupted sync transfer will not resume in `scr_poststage.sh` since the source file is gone (`AXL_Resume()` returns -1).  The bad news is that I removed the return code from resume_transfer() in review, so `scr_poststage.sh` wasn't seeing the failure.  Once I restored the return code, `scr_poststage.sh` correctly caught the error and didn't finalize the dataset.  I'll include the fix in my next push.

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+			continue
+		fi
+
+		echo "Setting current dataset to $DSET" >> $LOGFILE
+		if ! scr_index -p $PREFIX --current=$DSET &>> $LOGFILE ; then
+			echo "Couldn't set current dataset to $DSET" >> $LOGFILE
+			continue
+		fi
```
I took a look at the code.  I think ``scr_index`` will avoid setting a non-checkpoint to be current.  However, it also will print a message and return with an error:

https://github.com/LLNL/scr/blob/55c800d60d621d538c48d1db2cc692a07477273a/src/scr_index.c#L2022

Let's avoid printing any error messages so that we don't throw a false alarm for the user.  You can use ``test_api`` to write a non-checkpoint by adding an ``--output`` option.  I think ``--output 1`` will write every dataset as pure output (non-checkpoint).

adammoody commented on src/scr_flush_file.c:
```diff
@@ -67,6 +252,11 @@ struct arglist {
   int location;   /* return the location of dataset with specified id in cache */
   int name;       /* dataset name (label) */
+  int summary;    /* Generate a summary file for a dataset.  This is useful
+                   * when you've manually transferred the checkpoints files
```
```suggestion
                   * when you've manually transferred the dataset files
```

adammoody commented on src/scr_flush_file.c:
```diff
@@ -67,6 +252,11 @@ struct arglist {
   int name;       /* dataset name (label) */
+  int summary;    /* Generate a summary file for a dataset.  This is useful
+                   * when you've manually transferred the checkpoints files
+                   * outside of SCR, and need to tell SCR that there complete.
```
```suggestion
                   * outside of SCR, and need to tell SCR that they are complete.
```

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,184 @@
+/*
+ * Remove a particular dataset from the flush file.
+ */
+static void
+scr_flush_file_remove_dataset(int id, char* flush_file_path)
+{
+    kvtree* hash = kvtree_new();
+    kvtree_read_file(flush_file_path, hash);
+
+    kvtree_unset_kv_int(hash, SCR_FLUSH_KEY_DATASET, id);
+
+    /* write the hash back to the flush file */
+    kvtree_write_file(flush_file_path, hash);
+
+    /* delete the hash */
+    kvtree_delete(&hash);
+}
```
Let's call ``scr_flush_file_dataset_remove`` in place of this.  It can move it to the "flush_nompi" file like the other functions.
https://github.com/LLNL/scr/blob/55c800d60d621d538c48d1db2cc692a07477273a/src/scr_flush_file_mpi.c#L83

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+	scr_index -l -p $PREFIX >> $LOGFILE
+
+	# Finalize each dataset listed in the flush file
+	for ID in $(scr_flush_file --dir $PREFIX --list-ckpt) ; do
```
Oh, I just realized there is a ``--list-ckpt`` here.  That does ensure this loop will only operate on checkpoint datasets, so the ``scr_index --current`` would be good.

However, we also want to finalize any pure output (non-checkpoint) datasets.  Refer to the ``scr_postrun`` script, which has a similar requirement but for scavenge operations.

adammoody commented:
Thanks @tonyhutter .  I took another pass at it.

adammoody commented on src/scr_util.c:
```diff
@@ -388,3 +388,56 @@ spath* scr_get_prefix(const char* str)
 }
+
+/* Clone of apsrintf().  See the standard asprintf() man page for details */
+int asprintf(char** strp, const char* fmt, ...)
```
It looks like all of the ``asprintf`` functions are for construction of path strings.  Let's drop ``asprintf`` and use ``spath`` for those instead.

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+
+# Please fill in BINDIR AND PREFIX below:
+
+# The full path to the dir containing the SCR binaries (scr_index and scr_flush_file)
+BINDIR=""
```
I think we could prefill this value for people by changing this to a ``.in`` file and processing it during the cmake configure step.
```
BINDIR="@X_BINDIR@"
```
We do this for other scripts like ``scr_postrun.in``:
https://github.com/LLNL/scr/blob/55c800d60d621d538c48d1db2cc692a07477273a/scripts/common/scr_postrun.in#L24

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+# SCR allows you to spawn off dataset transfers "in the background"
+# that will finish some time after a job completes.  This saves you from
+# using your compute time to transfer datasets.  You can do this by
+# specifying XFER=BBAPI_POSTSTAGE for a storage descriptor in scr.conf.
```
This comment is out of date now.

adammoody commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+#
+# This script is to be run as a 2nd-half post-stage script on an IBM
+# system.  A 2nd-half post-stage script will run after all the job's burst
+# buffers transfers have finished (which could be hours later after the job
```
```suggestion
# buffer transfers have finished (which could be hours later after the job
```

adammoody commented on src/scr_flush_file.c:
```diff
@@ -40,6 +45,184 @@
+
+  id = AXL_Create(AXL_XFER_STATE_FILE, "scr", state_file_path);
+  if (id < 0) {
+    scr_err("Error: AXL_Create() = %d\n", id);
```
For ``scr_err`` calls, let's change them to be of this form:
```suggestion
    scr_err("Error: AXL_Create() = %d @ %s:%d", id, __FILE__, __LINE__);
```
The function adds a trailing newline, so that's not needed.  We also tack on the file and line number to every message, because that info helps in debugging problems reported by users.

tonyhutter commented on scripts/scr_poststage.sh:
```diff
@@ -0,0 +1,90 @@
+	scr_index -l -p $PREFIX &>> $LOGFILE
+}
+
+LOCKFILE="$PREFIX/lockfile"
+# Make sure only one poststage is running this script at a time.  Wait a max of 10
+# seconds for the lock.
+(
+	flock -n 9 || exit 1
+	do_poststage
+) 9>$LOCKFILE
```
I think I put the comment in and forgot to add `--timeout 10`.  I'll leave it as-is so it waits forever.

tonyhutter replied:
... and remove the comment about the 10 seconds.

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+
+CKPT=0 INTERVAL=1 GROUP=NODE   STORE=$BB_DIR TYPE=PARTNER
+
+SCR_CNTL_BASE=$BB_DIR
```
Let's take this out, since we don't want to suggest that people use the BB for the control directory.
```suggestion
```

tonyhutter replied:
I tried commenting out that line, but got:
```
Failed to create store descriptor for control directory [/dev/shm] @ /g/g0/hutter2/scr_poststage/src/scr_storedesc.c:381
```

adammoody replied:
Dang!  That's coming from here:
https://github.com/LLNL/scr/blob/develop/src/scr.c#L736-#L753

We have the STORE descriptor processing set up for an all or nothing.

In that case, the best option is to add a ``STORE`` line for ``/dev/shm``.
```
STORE=/dev/shm GROUP=NODE
```

tonyhutter replied:
:+1:

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+# # Allocate one node with 100GB burst buffer, run test_scr_poststag to
+# # create and begin transferring checkpoints, and after the transfers finish
+# # (after the job ends), call scr_poststage to finalize them.
+# bsub -nnodes 1 -stage storage=100:out=,scr_poststage lexec jsrun -r 1 test_scr_poststage
```
Let's drop lexec if that still works, since lexec is specific to LLNL.
```suggestion
# bsub -nnodes 1 -stage storage=100:out=,scr_poststage jsrun -r 1 test_scr_poststage
```

adammoody commented on src/scr_flush_sync.c:
```diff
@@ -225,9 +231,6 @@ int scr_flush_sync(scr_cache_index* cindex, int id)
   /* free data structures */
   kvtree_delete(&file_list);
 
-  /* remove sync flushing marker from flush file */
-  scr_flush_file_location_unset(id, SCR_FLUSH_KEY_LOCATION_SYNC_FLUSHING);
-
```
I think we want to leave this in.  Do you have to remove it for some reason?

tonyhutter replied:
Good catch, that should be left in.  I think it's an artifact from the my old `XFER=SCR_POSTSTAGE` code.

adammoody commented:
Thanks, @tonyhutter .  There are two more subtle details we need to deal with correctly in scr_poststage:
1) A dataset can be marked as both ``SCR_FLAG_CHECKPOINT`` and ``SCR_FLAG_OUTPUT`` at the same time.  I think in this case, the new code will attempt to flush such a dataset twice.  We should avoid that if possible or at least verify that it works correctly when that happens.
2) We only want to advance the current marker to the most recent checkpoint such that we have recovered all output datasets.  For example, consider this scenario:
```
id = 1 checkpoint, finalized
id = 2 output, transfer failed
id = 3 checkpiont, finalized
```
In the above case, even though we successfully finalize dataset 3, we don't want to mark that checkpoint as "current", because we also failed to finalize an earlier output dataset (id=2).  In that case, we want to mark dataset 1 as current.  This way the application restarts from that checkpoint (id=2) and then rewrite its output dataset at id=2.  If the app instead restarted from checkpoint (id=3) they would be missing output dataset 2.

You can see how both of those cases are handled in the ``scr_postrun`` script.

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+# bsub -nnodes 1 -stage storage=100:out=,scr_poststage lexec jsrun -r 1 test_scr_poststage
+#
+
+# Where do you want to write your checkpoints to?  This should be somewhere in
+# on the burst buffer filesystem.
+
+BB_DIR="$BBPATH/my_checkpoints"
+
```
Let's simplify the test to use ``$BBPATH`` rather than some directory in ``$BBPATH`` since that's how we'll be recommending that users configure things.
```suggestion
```

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+SCR_CACHE_BYPASS=0
+
+# Make our burst buffer a cache device
+STORE=$BB_DIR GROUP=NODE COUNT=1 FLUSH=BBAPI
+
+CKPT=0 INTERVAL=1 GROUP=NODE   STORE=$BB_DIR TYPE=PARTNER
```
```suggestion
STORE=$BBPATH GROUP=NODE COUNT=1 FLUSH=BBAPI

CKPT=0 INTERVAL=1 GROUP=NODE STORE=$BBPATH TYPE=PARTNER
```

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+
+make_scr_conf > /tmp/test_scr_poststage.config
+
+mkdir -p $BB_DIR $GPFS_DIR
```
```suggestion
mkdir -p $GPFS_DIR
```

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+SCR_DEBUG=10"
+}
+
+make_scr_conf > /tmp/test_scr_poststage.config
```
Let's move the config file to GPFS since that's more in line with what we recommend to users.  This line will then need to come after the ``mkdir -p $GPFS_DIR`` line with that switch.
```suggestion
make_scr_conf > /$GPFS_DIR/test_scr_poststage.config
```

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+fi
+
+# Start the checkpoints
+SCR_CONF_FILE=/tmp/test_scr_poststage.config test_api --times=3 --size=1G --output=3
```
```suggestion
SCR_CONF_FILE=/tmp/test_scr_poststage.config jsrun -r 1 test_api --times=3 --size=1G --output=3
```

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,81 @@
+fi
+
+# Start the checkpoints
+SCR_CONF_FILE=/tmp/test_scr_poststage.config test_api --times=3 --size=1G --output=3
+
+# To simulate a job crashing while doing a sync flush, remove the
+# 'SCR_FLUSH_ASYNC=1' line, and uncomment the line below:
+# SCR_CONF_FILE=/tmp/test_scr_poststage.config timeout --foreground -s KILL 13 test_api --times=1 --size=1G
+# SCR_CONF_FILE=/tmp/test_scr_poststage.config timeout --foreground -s KILL 13 test_api --times=10 --size=100M
+
+# Another test you can run
+# SCR_CONF_FILE=/tmp/test_scr_poststage.config test_api_multiple 1GB 10 0
```
```suggestion
SCR_CONF_FILE=/tmp/test_scr_poststage.config jsrun -r 1 test_api --times=3 --size=1G --output=3

# To simulate a job crashing while doing a sync flush, remove the
# 'SCR_FLUSH_ASYNC=1' line, and uncomment the line below:
# SCR_CONF_FILE=/tmp/test_scr_poststage.config timeout --foreground -s KILL 13 jsrun -r 1 test_api --times=1 --size=1G
# SCR_CONF_FILE=/tmp/test_scr_poststage.config timeout --foreground -s KILL 13 jsrun -r 1 test_api --times=10 --size=100M

# Another test you can run
# SCR_CONF_FILE=/tmp/test_scr_poststage.config jsrun -r 1 test_api_multiple 1GB 10 0
```

tonyhutter commented:
@adammoody thanks for giving this another look.  I'll implement the changes.

tonyhutter commented:
> 1.  A dataset can be marked as both SCR_FLAG_CHECKPOINT and SCR_FLAG_OUTPUT at the same time. 

I gave it a test and it seems to work (extra debug printfs added to log):
```
Creating directory: /mnt/bb_a4d3cedc2ddc5f884004991479a27aab/hutter2/scr.defjobid/scr.dataset.3/.scr
Starting dataset 3 `ckpt.3'
scr_complete_output: 12.450995 secs, 1.717987e+10 bytes, 1315.878840 MB/s, 82.242428 MB/s per proc
scr_reddesc_apply: 2.758427 secs, 1.717987e+10 bytes, 5939.616750 MB/s, 371.226047 MB/s per proc
Completed dataset 3 with return code 0
async flush attempt (is_output 1, is_ckpt 1)@ /g/g0/hutter2/scr_poststage/src/scr.c:417
Initiating async flush of dataset 3 `ckpt.3'
Flush of dataset 3 is ongoing
Completed checkpoint 3.
...
Begin flushing checkpoints that haven't started @ /g/g0/hutter2/scr_poststage/src/scr.c:2383
...
```
In poststage mode, I believe the last log line is a NOP and it should just leave the ongoing transfer alone:
                               
```C
    if (scr_my_rank_world == 0) {
      scr_dbg(2, "Begin flushing checkpoints that haven't started @ %s:%d", __FILE__, __LINE__);
    }

    if (poststage) {
      if (scr_flush_file_is_flushing(scr_ckpt_dset_id)) {
        flush_rc = SCR_SUCCESS;
      } else {
        flush_rc = scr_flush_async_start(scr_cindex, scr_ckpt_dset_id);
      }
    } else {
      /* Normal case */
      flush_rc = scr_flush_sync(scr_cindex, scr_ckpt_dset_id);
    }
```

tonyhutter commented:
@adammoody I believe my latest push has all the changes you were looking for.  Note that I couldn't think of a good way to setup your failure example for testing:
```
id = 1 checkpoint, finalized
id = 2 output, transfer failed
id = 3 checkpoint, finalized
```
So I added the logic to deal with this (which looks correct), and it at least works for the common case, but I didn't test this specific case.

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,86 @@
+
+CKPT=0 INTERVAL=1 GROUP=NODE   STORE=$BBPATH TYPE=PARTNER
+
+SCR_CNTL_BASE=/dev/shm
```
Let's try to take this out one more time.  I merged a change that should now create a store descriptor for the control directory by default.  You will likely need to rebase on the latest develop to pick it up.

adammoody commented on examples/test_scr_poststage:
```diff
@@ -0,0 +1,86 @@
+# Make our burst buffer a cache device
+STORE=$BBPATH GROUP=NODE COUNT=1 FLUSH=BBAPI
+
+# Just need this entry for SCR_CNTL_BASE
+STORE=/dev/shm GROUP=NODE
```
These lines, too, as below.

adammoody commented:
We have a parallel set of CMake files used to build SCR in the "dist" directory.  For now, we need to maintain both CMake paths.  To test that, first create an scr-top tarball, which you can do with:
```
cd dist
./builddist develop
```
Then you can build that:
```
mkdir temp
cd temp
tar -zxf scr-top-develop.tgz
cd scr-top-develop
cmake ...
make -j install
```
Can you also please update those files so that it picks up your changes in this PR?

adammoody commented:
Thanks @tonyhutter .  I think those few items above are the last things to check.  Should be good to go after that.

tonyhutter commented:
> Can you also please update those files so that it picks up your changes in this PR?

I'm not sure I understand what needs to be updated.  Won't my PR just go into `develop` after it's merged and it will get picked up by the script?

adammoody commented:
> > Can you also please update those files so that it picks up your changes in this PR?
> 
> I'm not sure I understand what needs to be updated. Won't my PR just go into `develop` after it's merged and it will get picked up by the script?

I am referring to the new CMakeLists.txt file at https://github.com/LLNL/scr/blob/develop/dist/CMakeLists.txt to account for any files you've added.  That new CMakeLists.txt file partly reuses our existing CMakeList.txt files and it partly duplicates logic, like the list of source files to be compiled.  The steps above show how to test a build using that other CMakeLists.txt.  We need to test that this build also picks up your files.

tonyhutter commented:
>  That new CMakeLists.txt file partly reuses our existing CMakeList.txt files and it partly duplicates logic, like the list of source files to be compiled. 

Won't the source list be dependent on the tag we pass to `builddist`?   For example, `builddist develop` will include `scr_flush_nompi.c` but `builddist v3.0rc1` wont.

tonyhutter commented:
> Won't the source list be dependent on the tag we pass to builddist? For example, builddist develop will include scr_flush_nompi.c but builddist v3.0rc1 wont.

I suppose as long as you passed `builddist` the same git tag you have checked out, you'd be fine.  It would only be if you tried running `builddist v3.0rc1` from the `scr:develop` git branch that you'd run into problems.  I wonder if it's better to remove the tag argument to `builddist` and just have it build whatever the current branch is (since that what builddist's `CMakeList.txt` would match up with).

adammoody commented:
Ugh!  Yes, you're right! That thought had occurred to me at one point when I was putting this together, but I forgot about it already.

One option would be for us to keep different CMakeLists.txt files around that correspond to the different SCR releases and have builddist pick the appropriate file for the given version.  That sounds like a pain, and I think it would need to grow to cover the full set of cmake files, not just the top-level one.  Another option would be to skip that manual versioning and rely on git to maintain the history for us according to the SCR tags.  I need to think about that for a bit, but I'm leaning toward that second git tag-based option.

For now, let's update the dist/CMakeLists.txt file to just worry about building a tarball of the "develop" branch.  Being able to build a tarball of develop is helpful since we often give those out to users to test out bug fixes.

adammoody reviewed (approved):
Thanks @tonyhutter .  This looks good.  Thanks for working though all of those changes!
